File: Data Structures & Algorithms/eating-bananas/submission-3.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = max(piles)
        ans = right
        while left <= right:
            mid = (left + right) // 2

            if self.eat(mid, piles) <= h:
                ans = mid
                # although we already find a solution, continue search for a optimal solution
                right = mid -1
            else:
                left = mid + 1
        
        return ans

    def eat(self, val: int, piles: List[int]):
        sum = 0
        for pile in piles:
            time = math.ceil(pile / val)
            sum += time
        return sum

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-3.java
class Solution {
    public int maxProfit(int[] prices) {
        int left = 0;
        int right = left + 1;
        int maxProfit = 0;

        while (right <= prices.length - 1) {
            if (prices[right] < prices[left]) {
                left = right;
            } else {
                int profit = prices[right] - prices[left];
                if (maxProfit < profit) {
                    maxProfit = profit;
                }
            }
            right++;
        }
        return maxProfit;
    }
}
